### PR TITLE
Add ADS simulation

### PR DESCRIPTION
**Pull request by jschueller**
This integrates code from the otads module, see
M. Munoz Zuniga, J. Garnier, E. Remy and E. de Rocquigny,
"Adaptative Directional Stratification for controlled estimation of the
probability of a rare event"

### CODE REVIEW

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+/* Constructor with parameters */
+AdaptiveDirectionalSampling::AdaptiveDirectionalSampling(const Event & event,
+    const RootStrategy & rootStrategy,
+    const SamplingStrategy & samplingStrategy)
+  : Simulation(event)
+  , standardEvent_(StandardEvent(event))
+  , rootStrategy_(rootStrategy)
+  , samplingStrategy_(samplingStrategy)
+  , gamma_(NumericalPoint(2, 0.5))
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: Default value in ResourceMap?

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+  , partialStratification_(false)
+  , maximumStratificationDimension_(3)
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: Default value in ResourceMap?

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+{
+
+  // First, reset the convergence history
+  convergenceStrategy_.clear();
+
+  // dimension
+  const UnsignedInteger dimension = standardEvent_.getImplementation()->getFunction().getInputDimension();
+
+  // stratification dimension
+  UnsignedInteger d = dimension;
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: const

Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: const? duplicate of dimension?

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+
+  // stratified dimensions
+  Indices strataIndices(d);
+  strataIndices.fill();
+
+  // number of subdivisions: 2^d
+  UnsignedInteger m = (1 << d);
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: const?

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+  UnsignedInteger n = 0;
+
+  // for each step
+  for (UnsignedInteger l = 0; l < L; ++ l)
+  {
+    const NumericalPoint w0(m, 1.0 / m);
+
+    NumericalScalar probabilityEstimate = 0.0;
+    NumericalScalar w0SigmaSum = 0.0;
+    NumericalPoint sigma(m, 0.);
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: 0.0

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+      quadrantSampling.setQuadrantOrientation(quadrantOrientation_);
+      quadrantSampling.setStrataIndices(strataIndices);
+      DirectionalSampling directionalSampling (getEvent(), rootStrategy_, quadrantSampling);
+      directionalSampling.setMaximumOuterSampling (ni);
+      directionalSampling.setBlockSize (blockSize);
+      directionalSampling.run();
+      const SimulationResult result(directionalSampling.getResult());
+      const NumericalScalar pf = result.getProbabilityEstimate();
+
+      if (pf > 0.)
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: 0.0

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+
+    const NumericalScalar varianceEstimate = w0SigmaSum * w0SigmaSum / (gamma_[l] * n); // (33)
+
+    // update result
+    setResult(SimulationResult(getEvent(), probabilityEstimate, varianceEstimate, n, blockSize));
+
+    // update weights
+    for (UnsignedInteger i = 0; i < m; ++ i)
+    {
+      w[i] = (w0SigmaSum > 0.) ? w0[i] * sigma[i] / w0SigmaSum : 0.; // (29)
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: 0.0

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+    }
+
+    if ((l == 0) && partialStratification_)
+    {
+      T_ = NumericalPoint(dimension);
+      for (UnsignedInteger k = 0; k < dimension; ++ k)
+      {
+        for (UnsignedInteger i = 0; i < m; ++ i)
+          if ((1 << k) & i)
+            T_[k] += fabs(T0[k][i] - T1[k][i]);
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: std::abs

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+      }
+
+      // sort variables according to T statistic
+      Indices order(dimension);
+      order.fill();
+      for (SignedInteger i = dimension - 1; i >= 0; -- i)
+        for (SignedInteger j = 0; j < i; ++ j)
+          if (T_[order[j]] < T_[order[j + 1]])
+          {
+            UnsignedInteger swap = order[j];
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: const

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+        if (k < maximumStratificationDimension_)
+          strataIndices.add(order[k]);
+      }
+
+      // new stratification dimension d' < d
+      UnsignedInteger d2 = strataIndices.getSize();
+      UnsignedInteger m2 = 1 << d2;
+
+      // compute new weights using existing simulations
+      NumericalPoint w2(m2, 0.);
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: 0.0

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+      // retrieve the p' variables contributing the most
+      strataIndices = Indices();
+      for (UnsignedInteger k = 0; k < dimension; ++ k)
+      {
+        if (k < maximumStratificationDimension_)
+          strataIndices.add(order[k]);
+      }
+
+      // new stratification dimension d' < d
+      UnsignedInteger d2 = strataIndices.getSize();
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: const?

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+      UnsignedInteger m2 = 1 << d2;
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: const?

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+      }
+
+      // update some parameters
+      m = m2;
+      w = w2;
+      d = d2;
+
+
+    } // if partialStratification_
+
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: Why so many blank lines?

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+RootStrategy AdaptiveDirectionalSampling::getRootStrategy() const
+{
+  return rootStrategy_;
+}
+
+/* Sampling strategy */
+void AdaptiveDirectionalSampling::setSamplingStrategy(const SamplingStrategy & samplingStrategy)
+{
+  UnsignedInteger dimension = getEvent().getImplementation()->getAntecedent()->getDistribution().getDimension();
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: const

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+}
+
+SamplingStrategy AdaptiveDirectionalSampling::getSamplingStrategy() const
+{
+  return samplingStrategy_;
+}
+
+void AdaptiveDirectionalSampling::setGamma(const NumericalPoint& gamma)
+{
+  UnsignedInteger dimension = gamma.getDimension();
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: const

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+  if (dimension > 2) throw InvalidDimensionException(HERE) << "gamma dimension is " << dimension;
+  NumericalScalar sum = 0.;
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: 0.0

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+  for (UnsignedInteger i = 0; i < dimension; ++ i)
+  {
+    if (gamma[i] <= 0.) throw InvalidArgumentException(HERE) << "gamma values should be positive";
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: 0.0

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 18]
```diff
@@ -0,0 +1,327 @@
...
+    sum += gamma[i];
+  }
+  if (fabs(sum - 1.0) > 1e-6) throw InvalidArgumentException(HERE) << "gamma components do not sum to 1";
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: 1e-6 in ResourceMap

File: lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx [hunk 19]
```diff
@@ -0,0 +1,327 @@
...
+
+NumericalPoint AdaptiveDirectionalSampling::getGamma() const
+{
+  return gamma_;
+}
+
+
+void AdaptiveDirectionalSampling::setQuadrantOrientation(const OT::NumericalPoint& quadrantOrientation)
+{
+  UnsignedInteger dimension = getEvent().getImplementation()->getAntecedent()->getDimension();
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/AdaptiveDirectionalSampling.cxx]: const

File: lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+{
+  NumericalSample result(samplingStrategy_.generate());
+  const UnsignedInteger size = result.getSize();
+  const UnsignedInteger strataDimension = stataIndices_.getSize();
+
+  for ( UnsignedInteger i = 0; i < size; ++ i )
+  {
+    for ( UnsignedInteger j = 0; j < strataDimension; ++ j )
+    {
+      result[i][stataIndices_[j]] = fabs( result[i][stataIndices_[j]] );
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx]: std::abs

File: lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+  if ( quadrantOrientation_.getDimension() > 1 )
+  {
+    for ( UnsignedInteger i = 0; i < size; ++ i )
+    {
+      NumericalPoint inP(strataDimension);
+      for ( UnsignedInteger j = 0; j < strataDimension; ++ j )
+      {
+        inP[j] = result[i][stataIndices_[j]];
+      }
+      NumericalPoint rotP(R_ * inP);
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx]: const

File: lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
+        result[i][stataIndices_[j]] *= -1.0;
+      }
+    }
+  }
+
+  if ( quadrantOrientation_.getDimension() > 1 )
+  {
+    for ( UnsignedInteger i = 0; i < size; ++ i )
+    {
+      NumericalPoint inP(strataDimension);
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx]: const

File: lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx [hunk 23]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+  }
+  return result;
+}
+
+
+void QuadrantSampling::setQuadrantIndex(const UnsignedInteger quadrantIndex)
+{
+  UnsignedInteger m = (1 << getStrataIndices().getSize());
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx]: const

File: lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx [hunk 24]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+NumericalPoint QuadrantSampling::getQuadrantOrientation() const
+{
+  return quadrantOrientation_;
+}
+
+void QuadrantSampling::setStrataIndices(const Indices & strataIndices)
+{
+  UnsignedInteger m = (1 << strataIndices.getSize());
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx]: const

File: lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx [hunk 25]
```diff
@@ -0,0 +1,209 @@
...
+  return oss;
+}
+
+
+void QuadrantSampling::updateRotation()
+{
+
+  if ( quadrantOrientation_.getDimension() > 0 )
+  {
+    UnsignedInteger p = getStrataIndices().getSize();
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx]: const

File: lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx [hunk 26]
```diff
@@ -0,0 +1,209 @@
...
+    SquareMatrix H(p);
+    for ( UnsignedInteger i = 0; i < p; ++ i )
+      for ( UnsignedInteger j = 0; j < p; ++ j )
+        H(j, i) = f[i][j];
+
+    // P is the transformation G -> F
+    SquareMatrix P(p);
+    for ( UnsignedInteger j = 0; j < p; ++ j )
+    {
+      P(0, j) = 1. / sqrt(1.0 * p);
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx]: 1.0

File: lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx [hunk 27]
```diff
@@ -0,0 +1,209 @@
...
+    }
+    for ( UnsignedInteger k = 1; k < p; ++ k )
+    {
+      P(k, k) = (p - k) / sqrt(1.0 * (p - k) * (p - k + 1));
+      for ( UnsignedInteger i = 0; i < p; ++ i )
+      {
+        if ((i < 1) || (i > k))
+          P(k, i) -= 1. / sqrt(1.0 * (p - k) * (p - k + 1));
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx]: 1.0

File: lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx [hunk 28]
```diff
@@ -0,0 +1,209 @@
...
+      }
+    }
+
+    // R is the tranformation E -> G
+    R_ = P * H;
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx]: Use the new matrix multiply that allows to specify if a matrix needs to be transposed, as H may be transposed if you use the constructor based on the raw content of f

File: lib/test/t_AdaptiveDirectionalSampling_std.cxx [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+  /* We create a 'usual' RandomVector from the Distribution */
+  RandomVector vect(myDistribution);
+
+  /* We create a composite random vector */
+  RandomVector output(function, vect);
+
+  /* We create an Event from this RandomVector */
+  Event myEvent(output, Less(), -3.0);
+
+  UnsignedInteger n = static_cast<UnsignedInteger>(1e3);
```
Comment (regislebrun) [inline lib/test/t_AdaptiveDirectionalSampling_std.cxx]: strange. n = 1000 is more readable.

File: lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx (hunk 28 above)
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/Simulation/QuadrantSampling.cxx]: In fact if I use the constructor based on raw contents it already transposes it as NS and Matrix have opposite storage orders.

File: lib/src/Uncertainty/Algorithm/Simulation/openturns/QuadrantSampling.hxx [hunk 30]
```diff
@@ -0,0 +1,90 @@
...
+  UnsignedInteger quadrantIndex_;
+
+  // first quadrant bisector
+  NumericalPoint quadrantOrientation_;
+
+  // rotation matrix
+  SquareMatrix R_;
+
+  // stratification indices
+  Indices stataIndices_;
```
Comment (dbarbier) [inline lib/src/Uncertainty/Algorithm/Simulation/openturns/QuadrantSampling.hxx]: Typo: should be `strataIndices_`

**Comment by dbarbier** [comment]
+1

**Pull request merged**